File: src/main/java/org/nest/nestml/prettyprinter/LegacyExpressionPrinter.java
/*
 * LegacyExpressionPrinter.java
 *
 * This file is part of NEST.
 *
 * Copyright (C) 2004 The NEST Initiative
 *
 * NEST is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * NEST is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with NEST.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.nest.nestml.prettyprinter;

import org.nest.nestml._ast.ASTExpr;
import org.nest.nestml._ast.ASTFunctionCall;

/**
 * Created by ptraeder.
 * "Legacy" version of the expression printer that does not print units for literals
 */
public class LegacyExpressionPrinter extends ExpressionsPrettyPrinter{

  public LegacyExpressionPrinter() {
    super();
  }

  public LegacyExpressionPrinter(final IReferenceConverter referenceConverter) {
    super(referenceConverter);
  }

  @Override
  protected String doPrint(final ASTExpr expr)  {
    if (expr.getNumericLiteral().isPresent()) { // number
      return typesPrinter().prettyprint(expr.getNumericLiteral().get());
    }

    if (expr.isInf()) {
      return convertConstant("inf");
    }
    else if (expr.getStringLiteral().isPresent()) { // string
      return typesPrinter().prettyprint(expr.getStringLiteral().get());
    }
    else if (expr.getBooleanLiteral().isPresent()) { // boolean
      return typesPrinter().prettyprint(expr.getBooleanLiteral().get());
    }
    else if (expr.getVariable().isPresent()) { // var
      return convertVariableName(expr.getVariable().get());
    }
    else if (expr.getFunctionCall().isPresent()) { // function
      final ASTFunctionCall astFunctionCall = expr.getFunctionCall().get();
      return printFunctionCall(astFunctionCall);

    }
    else if (expr.isUnaryPlus()) {
      return "+" + print(expr.getTerm().get());
    }
    else if (expr.isUnaryMinus()) {
      return "-" + print(expr.getTerm().get());
    }
    else if (expr.isUnaryTilde()) {
      return "~" + print(expr.getTerm().get());
    }
    else if (expr.isLeftParentheses() && expr.isRightParentheses()) {
      return "(" + print(expr.getExpr().get()) + ")";
    }
    else if (expr.isPlusOp() || expr.isMinusOp() || expr.isTimesOp() || expr.isDivOp()) {
      final StringBuilder expression = new StringBuilder();
      final String leftOperand = print(expr.getLeft().get());
      final String rightOperand = print(expr.getRight().get());
      expression.append(leftOperand);
      expression.append(getArithmeticOperator(expr));
      expression.append(rightOperand);
      return expression.toString();
    }
    else if (expr.isPow()) {
      final String leftOperand = print(expr.getBase().get());
      final String rightOperand = print(expr.getExponent().get());

      final String powTemplate = referenceConverter.convertBinaryOperator("**");
      return String.format(powTemplate, leftOperand, rightOperand);
    }
    else if (expr.isShiftLeft() ||
        expr.isShiftRight() ||
        expr.isModuloOp() ||
        expr.isBitAnd() ||
        expr.isBitOr() ||
        expr.isBitXor()) {
      final StringBuilder expression = new StringBuilder();
      final String leftOperand = print(expr.getLeft().get());
      final String rightOperand = print(expr.getRight().get());
      expression.append(leftOperand);
      expression.append(printBitOperator(expr));
      expression.append(rightOperand);
      return expression.toString();
    }
    // left:Expr (lt:["<"] | le:["<="] | eq:["=="] | ne:["!="] | ne2:["<>"] | ge:[">="] | gt:[">"]) right:Expr
    else if (expr.isLt() ||
        expr.isLe() ||
        expr.isEq() ||
        expr.isNe() ||
        expr.isNe2() ||
        expr.isGe() ||
        expr.isGt()) {
      final StringBuilder expression = new StringBuilder();
      final String leftOperand = print(expr.getLeft().get());
      final String rightOperand = print(expr.getRight().get());
      expression.append(leftOperand).append(printComparisonOperator(expr)).append(rightOperand);
      return expression.toString();
    }
    else if (expr.isLogicalOr() || expr.isLogicalAnd()) {
      final String leftOperand = print(expr.getLeft().get());
      final String rightOperand = print(expr.getRight().get());

      if (expr.isLogicalAnd()) {
        final String operatorTemplate = referenceConverter.convertBinaryOperator("and");
        return String.format(operatorTemplate, leftOperand, rightOperand);
      }
      else { // it is an or-operator
        final String operatorTemplate = referenceConverter.convertBinaryOperator("or");
        return String.format(operatorTemplate, leftOperand, rightOperand);
      }

    }
    else if (expr.isLogicalNot()) {
      return "not " + print(expr.getExpr().get());
    }
    else if (expr.getCondition().isPresent()) {
      final String condition = print(expr.getCondition().get());
      final String ifTrue = print(expr.getIfTrue().get()); // guaranteed by grammar
      final String ifNot = print(expr.getIfNot().get()); // guaranteed by grammar
      return "(" + condition + ")?(" + ifTrue + "):(" + ifNot + ")";
    }

    final String errorMsg = "Unsupported grammar element:  PrettyPrinter must be fixed " + expr.get_SourcePositionStart();

    throw new RuntimeException(errorMsg);
  }
}

File: pynestml/codegeneration/legacy_expression_printer.py
#
# legacy_expression_printer.py
#
# This file is part of NEST.
#
# Copyright (C) 2004 The NEST Initiative
#
# NEST is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# NEST is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with NEST.  If not, see <http://www.gnu.org/licenses/>.
from pynestml.codegeneration.expressions_pretty_printer import ExpressionsPrettyPrinter
from pynestml.codegeneration.i_reference_converter import IReferenceConverter
from pynestml.codegeneration.idempotent_reference_converter import IdempotentReferenceConverter
from pynestml.meta_model.ast_expression import ASTExpression
from pynestml.meta_model.ast_simple_expression import ASTSimpleExpression


class LegacyExpressionPrinter(ExpressionsPrettyPrinter):
    """
    An adjusted version of the pretty printer which does not print units with literals.
    """

    def __init__(self, reference_converter=None):
        """
        Standard constructor.
        :param reference_converter: a single reference converter object.
        :type reference_converter: IReferenceConverter
        """
        from pynestml.codegeneration.expressions_pretty_printer import TypesPrinter
        super(LegacyExpressionPrinter, self).__init__(reference_converter)
        if reference_converter is not None:
            self.reference_converter = reference_converter
        else:
            self.reference_converter = IdempotentReferenceConverter()
        self.types_printer = TypesPrinter()

    def do_print(self, node):
        """
        Prints a single rhs.
        :param node: a single rhs.
        :type node: ASTExpression or ASTSimpleExpression.
        :return: string representation of the rhs
        :rtype: str
        """
        # todo : printing of literals etc. should be done by constant converter, not a type converter
        if isinstance(node, ASTSimpleExpression):
            if node.is_numeric_literal():
                return self.types_printer.pretty_print(node.get_numeric_literal())
            elif node.is_inf_literal:
                return self.reference_converter.convert_constant('inf')
            elif node.is_string():
                return self.types_printer.pretty_print(node.get_string())
            elif node.is_boolean_true:
                return self.types_printer.pretty_print(True)
            elif node.is_boolean_false:
                return self.types_printer.pretty_print(False)
            elif node.is_variable():
                return self.reference_converter.convert_name_reference(node.get_variable())
            elif node.is_function_call():
                return self.print_function_call(node.get_function_call())
        elif isinstance(node, ASTExpression):
            if node.is_unary_operator():
                op = self.reference_converter.convert_unary_op(node.get_unary_operator())
                rhs = self.print_expression(node.get_expression())
                return op % rhs
            elif node.is_encapsulated:
                return self.reference_converter.convert_encapsulated() % self.print_expression(node.get_expression())
            # logical not
            elif node.is_logical_not:
                op = self.reference_converter.convert_logical_not()
                rhs = self.print_expression(node.get_expression())
                return op % rhs
            # compound rhs with lhs + rhs
            elif node.is_compound_expression():
                lhs = self.print_expression(node.get_lhs())
                op = self.reference_converter.convert_binary_op(node.get_binary_operator())
                rhs = self.print_expression(node.get_rhs())
                return op % (lhs, rhs)
            elif node.is_ternary_operator():
                condition = self.print_expression(node.get_condition())
                if_true = self.print_expression(node.get_if_true())
                if_not = self.print_expression(node.if_not)
                return self.reference_converter.convert_ternary_operator() % (condition, if_true, if_not)
        else:
            raise RuntimeError('Unsupported rhs in rhs pretty printer (%s)!' % str(node))
